Add gpio nodes to stm32 based soc/boards

erwango opened the pull request:
Aim of this pull request is to provide gpio related nodes to stm32 socs and boards.

### Provide STM32 GPIO node ddescriptions in dts files
- gpio nodes are populated on STM32 SoC dtsi files
- gpio related nodes (button, led) are provided in boards dts files. For instance a node will be described in dts as follows:
```
	leds {
		compatible = "gpio-leds";
		green_led_0: led@0 {
			gpios = <&gpioc 13 GPIO_INT_ACTIVE_HIGH>;
			label = "USR0 LED";
		};
```
- In order to enable them,  dt and yaml bindings are provided.
- exti nodes, as gpio interrupt parents are introduced.

### GPIO generation
- GPIO 'port' is renamed 'controller' to fit dts description
- Generate GPIO related definitions, for instance on LED node described above
```
#define GPIO_LEDS_0_GPIO_CONTROLLER	"GPIOE"
#define GPIO_LEDS_0_GPIO_FLAGS_0	4
#define GPIO_LEDS_0_GPIO_PIN_0		0
#define GPIO_LEDS_0_LABEL		"User LD1"
```
- Generate node aliases for generic use in sample applications. For LED example, this will be:
```
#define LED0_GPIO_CONTROLLER		GPIO_LEDS_0_GPIO_CONTROLLER
#define LED0_GPIO_FLAGS			GPIO_LEDS_0_GPIO_FLAGS
#define LED0_GPIO_PIN			GPIO_LEDS_0_GPIO_PIN
```

### Applications impact
Zephyr generic LED0_GPIO_PORT and SW0_GPIO_NAME are reworked to generic version:
LED0_GPIO_CONTROLLER and SW0_GPIO_CONTROLLER
- Samples and tests application using these are update to support new alternate version. Once all boards updated vs this new GPIO controller 
- STM32 boards board.h files are cleaned up
- Tested on button and blinky samples

EDIT (12/19): Led and keys yaml bindings added 
EDIT (12/19): Added GPIOs definition generation, changing the scope of the RFC. @superna9999, @ydamigos: you might want have a new look and update your review comment
EDIT (04/11): Description updated, PR ready for complete review

b0661 commented on dts/arm/st/stm32f091.dtsi:
```diff
@@ -5,3 +5,19 @@
+	soc {
+		pinctrl: pin-controller {
+
+			gpioe: gpio@48001000 {
```
Is there any difference to the gpioe already defined in stm32f0.dtsi?

erwango replied:
good catch, thanks. This one does not belong here.

erwango replied:
To be exact:
This one is correct, wrong one was gpioe in stm32f0.dtsi and it is now corrected

b0661 commented on dts/arm/st/stm32f0.dtsi:
```diff
@@ -49,6 +49,60 @@
 			#size-cells = <1>;
 			reg = <0x48000000 0x1800>;
+
+			gpioa: gpio@48000000 {
```
Is there a special reason to put gpio nodes within the pinctrl node besides that it is done in linux?

This just makes the defines longer or needs some extra logic to strip off the extra "PINCTRL" label.
A gpio device may just use the pinctrl/pinmux device but can be seen independent from the pinmux/pinctrl device.

b0661 commented on dts/arm/st/stm32f0.dtsi:
```diff
@@ -49,6 +49,60 @@
 			#size-cells = <1>;
 			reg = <0x48000000 0x1800>;
+
+			gpioa: gpio@48000000 {
```
Is there a special reason to put gpio nodes within the pinctrl node besides that it is done in linux?

This just makes the defines longer or needs some extra logic to strip off the extra "PINCTRL" label.
A gpio device may just use the pinctrl/pinmux device but can be seen independent from the pinmux/pinctrl device.

erwango replied:
One reason is indeed to stay close to Linux.
Other reason is that on stm32 MCUS, gpio and pin control is done at in same IP. GPIO and PIN only differ from alternate function definition.
Then, we can argue that this should be done the other way round and pinctrl might better be inside gpio node. 
Anyway, this is something that we might have to rework inside stm32 zephyr drivers for a better match.  But this was not the purpose here.

ydamigos commented on dts/arm/st/stm32f405.dtsi:
```diff
@@ -10,7 +10,34 @@
+				#gpio-cells = <2>;
+				reg = <0x40021800 0x400>;
+				clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x00000040>;
+				label = "GPIOH";
```
GPIOG?

ydamigos replied:
This is correct.

ydamigos commented on dts/arm/st/stm32f412.dtsi:
```diff
@@ -9,6 +9,28 @@
+				#gpio-cells = <2>;
+				reg = <0x40021800 0x400>;
+				clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x00000040>;
+				label = "GPIOH";
```
GPIOG?

erwango replied:
Right, thanks!

ydamigos commented on dts/arm/st/stm32f3.dtsi:
```diff
@@ -46,6 +46,48 @@
+				label = "GPIOD";
+			};
+
+			gpiof: gpio@480021000 {
```
0x48001400?

erwango replied:
Right, thanks!

ydamigos commented on dts/arm/st/stm32f0.dtsi:
```diff
@@ -49,6 +49,51 @@
+				label = "GPIOD";
+			};
+
+			gpiof: gpio@480021000 {
```
0x48001400?

erwango replied:
Right, thanks!

ydamigos reviewed (dismissed):
Some minor issues. LGTM.

mbolivar commented on boards/arm/96b_carbon/96b_carbon.dts:
```diff
@@ -16,6 +16,37 @@
 	};
+
+	leds {
+		compatible = "gpio-leds";
```
I'm still learning about how Zephyr uses DT, but I can't find a property in dts/bindings that has a compatible constraint "gpio-leds". What are the rules here; does this need to be written too or can compatible properties be set in dts files even if they don't have accompanying YAML bindings in Zephyr?

I've looked at extract_dts_includes.py and I've read http://docs.zephyrproject.org/devices/dts/device_tree.html, but I'm still confused.

erwango replied:
@mbolivar: You're right, a yaml binding file is required to extract information.
Besides, some script changes are also required to get #defines generation working. I'm working on it right now.
Very next step would be to be able to generate full code as proposed for pinmux (#5021). This whole task might take some while, then I'd like to be able to split it in two and having this first dts part merged while I complete generation part. Does its sound ok to you?

erwango replied:
@mbolivar:Thanks for the feedback. FYI, in the end I updated scope of current RFC to cover definitions generation as well.

erwango committed:
scripts: extract_dts_inlcudes: look up compatible field in parents

Modify get_compat function to look for 'compatible' property
in parent nodes if not available at node level.
Since this operation is quite common, this allows some code
factorization. As a consequence, get_compat takes 'node_address'
as argument instead of 'node'.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on include/dt-bindings/gpio/gpio.h:
```diff
@@ -0,0 +1,158 @@
+/*
```
> If we cannot source gpio.h within DT system

Indeed, dt parser is not able to parse include files with structs or enum. So it requires separate 
include files limited to #define.

> maybe we need to split it in two files but keep both of them directly in include/ folder?

I had the same doubts but I found it better to all al dt includes in dt-bindings. One way or the other, consistency could be discussed.

> In any case we have to make sure we don't break generation of doxygen documentation.

Agreed, did I broke something yet? I'll have a check. Thanks for the comment

nashif replied:
agree with @mnkp , if you are going to split the header, then please keep it in include/ along-side gpio.h, putting this in a dt specific location is not consistent and confusing.

erwango replied:
@nashif, @mnkp  : The other reason why I made this choice is this is the usage in device tree in general: device tree is supposed to be self-sufficient/self contained. 
I haven't been able to find this documented somewhere, but one can check and confirm on Linux device tree files for instance. 
Besides long term target of having one single repo for device tree will also require this property.
Maybe @galak can comment on this.

galak replied:
Some splitting of the header is going to be required.  In practice users should only be specifying these GPIO params in the device tree file.

erwango committed:
scripts: extract_dts_includes: generate controller #define's

When a node is referencing to a parent node and this node is
reported as controller, generate a #define to declare this controller.
Value is controller property 'label'
For instance, if following gpio controller is referenced in board dts:
green_led_1: led@1 {
    gpios = <&gpioa 5 GPIO_INT_ACTIVE_HIGH>;
    label = "User LD1";
};
Following will be generated:
\#define GPIO_LEDS_1_GPIO_CONTROLLER "GPIOA"

Besides, if defined, alias label will be generated as well for this
controller:
aliases {
    led0 = &green_led_1;
};
will trigger generation of:
\#define LED0_GPIO_CONTROLLER GPIO_LEDS_1_GPIO_CONTROLLER

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

mnkp commented on boards/arm/96b_carbon/96b_carbon.dts:
```diff
@@ -16,6 +16,37 @@
+
+	leds {
+		compatible = "gpio-leds";
+		green_led_1: led@0 {
```
Since we provide a dedicated device node for LED shouldn't we also provide a dedicated device driver, i.e. LED API? The current approach is somewhat inconsistent.

erwango replied:
Proposed LED objects could be driven with gpio driver.
Proposed compatible/yaml allow to generate information that could be used with GPIO driver.
I'm not sure a LED driver is required for current purpose. But I expect an upcoming LED driver could be based on this.
In the opposite case, a specific compatible could be defined with the new driver.

superna9999 commented on boards/arm/96b_carbon/96b_carbon.dts:
```diff
@@ -16,6 +16,37 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+		user_button: button@0 {
+			label = "User";
```
The label should reflect the board marking/schematics.

erwango replied:
Indeed, label should reflect board slikscreen and/or schematics.
In current case, I think this is a copy/paste issue as there is no "User" button on Carbon board.
I'll remove this one

erwango commented on boards/arm/96b_carbon/96b_carbon.dts:
```diff
@@ -16,6 +16,37 @@
 	};
+
+	leds {
+		compatible = "gpio-leds";
```
I actually adapted this from Linux on which this approach seems widely used (not only on ST chips).
About inheritance, while it is not mentioned in device tree spec, it appears to be a default rule.

erwango commented:
Rebased and addressed @galak comments.

@b0661, @tbursztyka:  I'm pushing this PR again since I think there are some parts worthy.
I don't see it as an alternative to #5799. For some parts it is an intermediate step and some others it is
complementary.
Let's see how we can use it/what can be used from it to move things forward.

Interesting parts from my point of view:
-Provides gpio node controllers on full stm32 range + matching yaml bindings
-Introduces zephyr aliases and reuse of generated bits directly into samples

tbursztyka commented on scripts/dts/extract/globals.py:
```diff
@@ -6,6 +6,7 @@
 #
 
 # globals
+compatibles = {}
```
left over from rebase?

erwango replied:
yep, removed, thanks

b0661 commented:
@erwango, @tbursztyka 

I think to be able to combine the different complementary solutions one more infrastructure patch is needed. Ideas:

- Base class for Device Tree Directives that the global functions. for define generation
  -  convert_string_to_label
  - get_node_label
  - get_label_string
  - get_node_label_string
- Device tree class/ singleton to bundle all access to device tree information and the
   global functions related to device tree.
  - get_all_compatibles
  - get_aliases
  - get_compat
  - get_chosen
  - get_phandles
  - get_reduced
  - find_parent_prop
- Defines class to bundle the defines info
  - insert_defs

erwango commented:
@b0661 :A priori Fine with me.

erwango commented:
recheck

erwango committed:
samples: Allow use of "CONTROLLER" postfix for LED and GPIO

Rework samples using LED and GPIO to enforce use of "CONTROLLER"
as LED and PGIO postifx.
Change impact all samples that could be run by boards yet moved to
GPIO node generations, which is only STM32 for now

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32 remove led and button definitions from board.h

Since they are now generated from dts, remove these definitions

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@nashif , @galak :
I have modified the commit impacting all the boards to limit to the impact on the boards directly concerned by the change. It allows to pass CI and transition to new names will be smoother.
Additionally, I had to tune one test and one board in order to avoid RAM overflow

erwango commented on samples/basic/button/src/main.c:
```diff
@@ -33,7 +33,7 @@
  * Change this to use a different interrupt trigger
  */
-#define EDGE    (GPIO_INT_EDGE | GPIO_INT_ACTIVE_LOW)
+#define EDGE    (GPIO_INT_EDGE | GPIO_INT_ACTIVE_HIGH)
```
Indeed, should have been cleaned-up. Fixed now, thanks!

erwango commented:
CI failed for unrelated issue:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/samples/net/coap_client/src/coap-client.c:179:6: error: ‘mgmt_event’ undeclared (first use in this function)
  if (mgmt_event != NET_EVENT_IF_UP) {
      ^~~~~~~~~~

jukkar commented:
> CI failed for unrelated issue:
> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/samples/net/coap_client/src/coap-client.c:179:6: error: ‘mgmt_event’ undeclared (first use in this function)
> if (mgmt_event != NET_EVENT_IF_UP) {
> ^~~~~~~~~~

I just sent a fix for this at #7037

galak commented:
Can you pull the fix for CI - tests/net into a separate PR

galak commented on dts/common/skeleton.dtsi:
```diff
@@ -9,4 +9,6 @@
 	#size-cells = <1>;
 	chosen { };
 	aliases { };
+	leds { };
```
Can you explain this further, why do we need empty nodes.  It's not typical to add these in skeleton.dtsi

erwango replied:
Was needed initially, but somehow it has became superfluous now. Removing the commit.

galak commented on dts/arm/st/stm32f0.dtsi:
```diff
@@ -55,11 +56,65 @@
 			label = "STM32_CLK_RCC";
 		};
 
+		exti: interrupt-controller@40010400 {
```
Do we have a yaml binding for exti?

galak commented on samples/basic/blinky/src/main.c:
```diff
@@ -10,7 +10,11 @@
 #include <gpio.h>
 
 /* Change this if you have an LED connected to a custom port */
+#ifdef LED0_GPIO_CONTROLLER
```
Rather than supporting both, we should make LED0_GPIO_CONTROLLER the default, and have this be something like:

#ifndef LED0_GPIO_CONTROLLER
#define LED0_GPIO_CONTROLLER LED0_GPIO_PORT
#endif

#define LED_PORT LED0_GPIO_CONTROLLER

galak commented on dts/bindings/gpio/gpio-leds.yaml:
```diff
@@ -0,0 +1,32 @@
+      constraint: "gpio-leds"
+
+    gpios:
+      type: string
```
Should this be compound?

galak commented on dts/bindings/gpio/gpio-keys.yaml:
```diff
@@ -0,0 +1,32 @@
+      constraint: "gpio-keys"
+
+    gpios:
+      type: string
```
should this be compound?

erwango replied:
ack

galak reviewed (commented):
The script change to lookup compat in parent, what is that needed for?

galak reviewed (commented):
Can you fix the commit message associated with ‘scripts: extract_dts_includes: generate controller #define’s’. It has a section that says ‘Following will be generated:’ but is empty.

erwango commented:
> The script change to lookup compat in parent, what is that needed for?

For a node that defines no compat , for instance 
```
'/gpio_keys/button@0' 
     {'type': <class 'type'>, 
      'name': 'button@0',
      'label': 'user_button',
      'props': {'gpios': [6, 13, 0], 'label': 'User'}, 
      'addr': 0}
```
we look up 'compatible' in parents properties:
```
'/gpio_keys' 
       {'type': <class 'type'>,
         'name': 'gpio_keys',
         'label': None,
         'props': {'compatible': 'gpio-keys', '#address-cells': 1, '#size-cells': 0},
         'addr': None}
```
This operation is quite common, so I have factorized the parent lookup inside get_compat function.
As a consequence, functions takes `node_address` as an argument, instead of `reduced[node_address]`.
I'm updating function header to make this clearer.
